internal/sqld/sqldwebsockets/websockets.go

```diff
@@ -0,0 +1,230 @@
+package sqldwebsockets
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strconv"
+
+	"nhooyr.io/websocket"
+	"nhooyr.io/websocket/wsjson"
+	"vitess.io/vitess/go/pools"
+)
+
+func errorMsg(errorResp interface{}) string {
+	return errorResp.(map[string]interface{})["error"].(map[string]interface{})["message"].(string)
+}
+
+func isErrorResp(resp interface{}) bool {
+	return resp.(map[string]interface{})["type"] == "response_error"
+}
+
+type SqldWebsocket struct {
+	conn   *websocket.Conn
+	idPool *pools.IDPool
+}
+
+type Params struct {
+	Names  []string
+	Values []any
+}
+
+func convertValue(v any) (map[string]interface{}, error) {
+	res := map[string]interface{}{}
+	if v == nil {
+		res["type"] = "null"
+	} else if integer, ok := v.(int64); ok {
+		res["type"] = "integer"
+		res["value"] = strconv.FormatInt(integer, 10)
+	} else if text, ok := v.(string); ok {
+		res["type"] = "text"
+		res["value"] = text
+	} else if blob, ok := v.([]byte); ok {
+		res["type"] = "blob"
+		res["value"] = base64.StdEncoding.EncodeToString(blob)
+	} else if float, ok := v.(float64); ok {
+		res["type"] = "float"
+		res["value"] = float
+	} else {
+		return nil, fmt.Errorf("unsupported value type: %s", v)
+	}
+	return res, nil
+}
+
+type ExecResponse struct {
+	resp map[string]interface{}
+}
+
+func (r *ExecResponse) AffectedRowCount() int64 {
+	return int64(r.resp["affected_row_count"].(float64))
+}
+
+func (r *ExecResponse) Columns() []string {
+	res := []string{}
+	cols := r.resp["cols"].([]interface{})
+	for idx := range cols {
+		var v string = ""
+		if cols[idx].(map[string]interface{})["name"] != nil {
+			v = cols[idx].(map[string]interface{})["name"].(string)
+		}
+		res = append(res, v)
+	}
+	return res
+}
+
+func (r *ExecResponse) RowsCount() int {
+	return len(r.resp["rows"].([]interface{}))
+}
+
+func (r *ExecResponse) RowLen(rowIdx int) int {
+	return len(r.resp["rows"].([]interface{})[rowIdx].([]interface{}))
+}
+
+func (r *ExecResponse) Value(rowIdx int, colIdx int) (any, error) {
+	val := r.resp["rows"].([]interface{})[rowIdx].([]interface{})[colIdx].(map[string]interface{})
+	switch val["type"] {
+	case "null":
+		return nil, nil
+	case "integer":
+		v, err := strconv.ParseInt(val["value"].(string), 10, 64)
+		if err != nil {
+			return nil, err
+		}
+		return v, nil
+	case "text":
+		return val["value"].(string), nil
+	case "blob":
+		base64Encoded := val["value"].(string)
+		v, err := base64.StdEncoding.DecodeString(base64Encoded)
+		if err != nil {
+			return nil, err
+		}
+		return v, nil
+	case "float":
+		return val["value"].(float64), nil
+	}
+	return nil, nil
```

Comment:
Perhaps the function should return an error in this case, instead of treating unrecognized values as NULLs?

Reply:
Yes. I wanted to do it like that but must have forgotten. Thanks.